common/src/main/java/com/wynntils/features/chat/ChatTabsFeature.java

```diff
@@ -125,13 +128,38 @@ public void onScreenRender(ScreenRenderEvent event) {
 
     @SubscribeEvent
     public void onChatScreenKeyTyped(ChatScreenKeyTypedEvent event) {
-        // We can't use keybinds here to not conflict with TAB key's other behaviours.
-        if (event.getKeyCode() != GLFW.GLFW_KEY_TAB) return;
         if (!(McUtils.mc().screen instanceof ChatScreen)) return;
-        if (!KeyboardUtils.isShiftDown()) return;
 
-        event.setCanceled(true);
-        Services.ChatTab.setFocusedTab(Services.ChatTab.getNextFocusedTab());
+        int keyCode = event.getKeyCode();
+        if (keyCode == GLFW.GLFW_KEY_TAB) {
+            boolean reverse = false;
+
+            if (KeyboardUtils.isShiftDown()) {
+                if (oldTabHotkey.get()) {
+                    Services.ChatTab.setFocusedTab(Services.ChatTab.getNextFocusedTab());
+                    event.setCanceled(true);
+                    return;
+                } else {
+                    reverse = true;
+                }
+            }
+
+            if (KeyboardUtils.isControlDown() && !oldTabHotkey.get()) {
+                int newTab = reverse ? Services.ChatTab.getLastFocusedTab() : Services.ChatTab.getNextFocusedTab();
+                Services.ChatTab.setFocusedTab(newTab);
+                event.setCanceled(true);
+                return;
+            }
```

Comment:
I think this could be simplified a bit, so the logic is more easily readable.

```
let newTab = -1;
if (oldTabHotkey) {
   if (shiftDown) {
      newTab = nextTab;
   }
} else {
   if (controlDown) {
      newTab = shiftDown ? lastTab : nextTab;   
   }
}

if (newTab == -1) return;

setFocusedTab + event.cancel
```

What do you think?

Reply:
That's better, I thought it was pretty ugly myself but I was out of ideas